File: packages/core/src/plugins/builtin/codex/history.ts
// packages/core/src/plugins/builtin/codex/history.ts
import type { ChatMessage, MessageContent } from '@qlan-ro/mainframe-types';
import type { ThreadItem, PatchChangeKind } from './types.js';
import type { CollabAgentToolCallItem } from './item-types.js';
import { parseUnifiedDiff } from '../../../messages/parse-unified-diff.js';
import { describeAgent, agentTitle, type AgentMetadata } from './thread-registry.js';

export function convertThreadItems(
  items: ThreadItem[],
  chatId: string,
  childItemsByThread: Map<string, ThreadItem[]> = new Map(),
  agentMetaByThread: Map<string, AgentMetadata> = new Map(),
): ChatMessage[] {
  const messages: ChatMessage[] = [];
  // Stash spawnAgent prompts (keyed by child thread id) so the matching `wait`
  // item can use them as the TaskGroup card's description.
  const spawnPrompts = new Map<string, string>();

  for (const item of items) {
    switch (item.type) {
      case 'agentMessage':
        messages.push(makeMessage(item.id, chatId, 'assistant', [{ type: 'text', text: item.text }]));
        break;

      case 'reasoning':
        messages.push(
          makeMessage(item.id, chatId, 'assistant', [
            { type: 'thinking', thinking: item.summary.join('\n') || item.content.join('\n') },
          ]),
        );
        break;

      case 'commandExecution':
        messages.push(
          makeMessage(item.id, chatId, 'assistant', [
            {
              type: 'tool_use',
              id: item.id,
              name: 'Bash',
              input: { command: item.command },
            },
          ]),
        );
        messages.push(
          makeMessage(`${item.id}:result`, chatId, 'tool_result', [
            {
              type: 'tool_result',
              toolUseId: item.id,
              content: item.aggregatedOutput ?? '',
              isError: item.exitCode !== undefined && item.exitCode !== 0,
            },
          ]),
        );
        break;

      case 'fileChange': {
        const isError = item.status === 'failed' || item.status === 'declined';
        for (const [index, change] of item.changes.entries()) {
          const toolId = `${item.id}:${index}`;
          const isAdd = change.kind.type === 'add';
          const toolName = isAdd ? 'Write' : 'Edit';
          const structuredPatch = parseUnifiedDiff(change.diff);
          const input: Record<string, unknown> = isAdd
            ? { file_path: change.path, content: extractAddedContent(change.diff) }
            : {
                file_path: change.path,
                old_string: '',
                new_string: '',
                ...((change.kind as Extract<PatchChangeKind, { type: 'update' }>).move_path != null
                  ? { move_path: (change.kind as Extract<PatchChangeKind, { type: 'update' }>).move_path }
                  : {}),
              };
          messages.push(
            makeMessage(toolId, chatId, 'assistant', [{ type: 'tool_use', id: toolId, name: toolName, input }]),
          );
          messages.push(
            makeMessage(`${toolId}:result`, chatId, 'tool_result', [
              {
                type: 'tool_result',
                toolUseId: toolId,
                content: 'OK',
                isError,
                ...(structuredPatch.length ? { structuredPatch } : {}),
              },
            ]),
          );
        }
        break;
      }

      case 'mcpToolCall': {
        const server = item.server ?? 'codex';
        const toolName = `mcp__${server}__${item.tool}`;
        messages.push(
          makeMessage(item.id, chatId, 'assistant', [
            {
              type: 'tool_use',
              id: item.id,
              name: toolName,
              input: item.arguments,
            },
          ]),
        );
        messages.push(
          makeMessage(`${item.id}:result`, chatId, 'tool_result', [
            {
              type: 'tool_result',
              toolUseId: item.id,
              content: item.error ? (item.error.message ?? '') : JSON.stringify(item.result?.content ?? ''),
              isError: !!item.error,
            },
          ]),
        );
        break;
      }

      case 'userMessage': {
        // Codex's `thread/read` returns `content: [{ type: 'text', text: '...' }]`.
        // Rollout JSONL records use `input_text` instead. Accept either, plus the
        // legacy top-level `item.text` field, so the same reader handles both shapes.
        const block = item.content?.find((b) => typeof b.text === 'string' && b.text.length > 0);
        const text = block?.text ?? item.text ?? '';
        if (!text) break;
        messages.push(makeMessage(item.id, chatId, 'user', [{ type: 'text', text }]));
        break;
      }

      case 'collabAgentToolCall': {
        // `spawnAgent` is dispatch metadata only — stash its prompt for the `wait` card.
        if (item.tool === 'spawnAgent') {
          for (const childId of item.receiverThreadIds ?? []) {
            if (item.prompt) spawnPrompts.set(childId, item.prompt);
          }
          break;
        }
        // `wait` renders the TaskGroup card with sub-agent's child items nested under it.
        emitCollabAgent(messages, chatId, item, spawnPrompts, childItemsByThread, agentMetaByThread);
        break;
      }

      // webSearch, todoList — skip for now
    }
  }

  return messages;
}

function emitCollabAgent(
  messages: ChatMessage[],
  chatId: string,
  item: CollabAgentToolCallItem,
  spawnPrompts: Map<string, string>,
  childItemsByThread: Map<string, ThreadItem[]>,
  agentMetaByThread: Map<string, AgentMetadata>,
): void {
  const isError = item.status === 'failed' || item.status === 'interrupted';
  const childId = item.receiverThreadIds?.[0];
  // Pull the agent's identity from Codex's thread DB:
  //   - subagent_type = nickname (e.g. "Maxwell") — bold card title, like Claude's
  //   - description   = the spawn prompt (the task) — informative subtitle, truncated
  //                     to 60 chars in the card with full text in a tooltip
  // If nickname is missing, fall back to role ("explorer") for the title.
  const meta = childId ? agentMetaByThread.get(childId) : undefined;
  const subagentType = agentTitle(meta) ?? describeAgent(meta) ?? 'Sub-agent';
  const prompt = (childId && spawnPrompts.get(childId)) ?? item.prompt ?? '';
  // Description is the agent's role (e.g. "explorer") — short subtitle next to the
  // nickname title. The full prompt is visible when the user expands the card.
  const description = describeAgent(meta) ?? (prompt || subagentType);
  const subAgentMessage = childId ? (item.agentsStates?.[childId]?.message ?? null) : null;

  // Build the parent assistant message: CollabAgent tool_use first, then sub-agent
  // child *non-result* blocks (tool_use, text, thinking) tagged with parentToolUseId so
  // the desktop's groupTaskChildren() nests them under it.
  //
  // Sub-agent tool_result blocks are NOT inlined here. groupMessages only attaches
  // results coming from separate `tool_result`-type messages — inlining them would lose
  // the bash output. We emit them as standalone tool_result messages right after the
  // parent so they get attached via toolUseId.
  const content: MessageContent[] = [
    {
      type: 'tool_use',
      id: item.id,
      name: 'CollabAgent',
      input: { prompt, description, subagent_type: subagentType },
    },
  ];
  const childToolResults: Array<MessageContent & { type: 'tool_result' }> = [];

  if (childId) {
    const childItems = childItemsByThread.get(childId);
    if (childItems && childItems.length > 0) {
      const childMessages = convertThreadItems(childItems, chatId, childItemsByThread);
      for (const m of childMessages) {
        // Skip the child thread's user-prompt echo — it's just a copy of the spawn prompt.
        if (m.type === 'user') continue;
        for (const block of m.content) {
          if (block.type === 'tool_result') {
            childToolResults.push({ ...block, parentToolUseId: item.id });
          } else {
            content.push({ ...block, parentToolUseId: item.id });
          }
        }
      }
    }
  }

  messages.push(makeMessage(item.id, chatId, 'assistant', content));
  for (const [index, r] of childToolResults.entries()) {
    messages.push(makeMessage(`${item.id}:child:${index}:result`, chatId, 'tool_result', [r]));
  }
  // Close the card with the CollabAgent's own tool_result (sub-agent's final message).
  messages.push(
    makeMessage(`${item.id}:result`, chatId, 'tool_result', [
      {
        type: 'tool_result',
        toolUseId: item.id,
        content: subAgentMessage ?? 'Sub-agent completed',
        isError,
      },
    ]),
  );
  if (childId) spawnPrompts.delete(childId);
}

/**
 * Build a ChatMessage with a CALLER-SUPPLIED deterministic id. The id is derived
 * from the Codex thread item's stable `id` (+ a slot suffix for items that emit
 * more than one message), so reconstructing the same items yields the same ids
 * every turn. That lets the display delta emitter detect appends/updates instead
 * of re-broadcasting the whole list (a `display.messages.set`) on every turn —
 * which previously happened because `id: nanoid()` changed the ids each pass.
 */
function makeMessage(id: string, chatId: string, type: ChatMessage['type'], content: MessageContent[]): ChatMessage {
  return {
    id,
    chatId,
    type,
    content,
    timestamp: new Date().toISOString(),
  };
}

function extractAddedContent(diff: string): string {
  return diff
    .split('\n')
    .filter((line) => line.startsWith('+') && !line.startsWith('+++'))
    .map((line) => line.slice(1))
    .join('\n');
}

File: packages/core-rs/crates/mainframe-adapter-codex/src/history.rs
//! Ported from `packages/core/src/plugins/builtin/codex/history.ts`.
//!
//! Hosts the crate-shared `MessageContent` block builders + the `with_parent`
//! tagger (reused by `event_mapper`); re-exports `convert_thread_items` (moved to
//! `history_convert.rs`, task 2, todo #247) and `parse_unified_diff` (moved to
//! `unified_diff.rs`) so external call sites keep compiling.

use std::collections::HashMap;

use mainframe_types::chat::{ChatMessage, ChatMessageType, DiffHunk, MessageContent};
use mainframe_types::content::LeafContent;
use serde_json::{Value, json};

use crate::item_types::PatchChangeKind;

pub use crate::history_convert::convert_thread_items;
pub(crate) use crate::unified_diff::parse_unified_diff;

/// Build a `ChatMessage` with a CALLER-SUPPLIED deterministic id (derived from the
/// Codex thread item's stable `id`), so reconstructing the same items yields the
/// same ids every turn (lets the display delta emitter detect appends/updates).
pub(crate) fn make_message(
    id: &str,
    chat_id: &str,
    r#type: ChatMessageType,
    content: Vec<MessageContent>,
) -> ChatMessage {
    ChatMessage {
        id: id.to_string(),
        chat_id: chat_id.to_string(),
        r#type,
        content,
        timestamp: mainframe_runtime::time::now_iso8601(),
        metadata: None,
    }
}

pub(crate) fn reasoning_text(summary: &[String], content: &[String]) -> String {
    let s = summary.join("\n");
    if s.is_empty() { content.join("\n") } else { s }
}

pub(crate) fn is_exec_error(exit_code: Option<i64>) -> bool {
    exit_code.map(|c| c != 0).unwrap_or(false)
}

pub(crate) fn bash_input(command: &str) -> HashMap<String, Value> {
    let mut m = HashMap::new();
    m.insert("command".to_string(), json!(command));
    m
}

pub(crate) fn file_change_input(
    is_add: bool,
    path: &str,
    diff: &str,
    kind: &PatchChangeKind,
) -> HashMap<String, Value> {
    let mut m = HashMap::new();
    m.insert("file_path".to_string(), json!(path));
    if is_add {
        m.insert("content".to_string(), json!(extract_added_content(diff)));
    } else {
        m.insert("old_string".to_string(), json!(""));
        m.insert("new_string".to_string(), json!(""));
        if let PatchChangeKind::Update {
            move_path: Some(mp),
        } = kind
        {
            m.insert("move_path".to_string(), json!(mp));
        }
    }
    m
}

pub(crate) fn mcp_result_content(
    content: Option<&Value>,
    error: Option<&crate::item_types::CodexItemError>,
) -> String {
    if let Some(err) = error {
        return err.message.clone();
    }
    // `JSON.stringify(item.result?.content ?? '')`
    let val = content
        .filter(|c| !c.is_null())
        .cloned()
        .unwrap_or(Value::String(String::new()));
    serde_json::to_string(&val).unwrap_or_default()
}

pub(crate) fn extract_added_content(diff: &str) -> String {
    diff.split('\n')
        .filter(|line| line.starts_with('+') && !line.starts_with("+++"))
        .map(|line| &line[1..])
        .collect::<Vec<_>>()
        .join("\n")
}

// ── Crate-shared MessageContent block builders (reused by event_mapper) ──────────

pub(crate) fn text_block(text: &str) -> MessageContent {
    MessageContent::Leaf(LeafContent::Text {
        text: text.to_string(),
        parent_tool_use_id: None,
    })
}

pub(crate) fn thinking_block(text: &str) -> MessageContent {
    MessageContent::Leaf(LeafContent::Thinking {
        thinking: text.to_string(),
        parent_tool_use_id: None,
    })
}

pub(crate) fn image_block(media_type: &str, data: &str) -> MessageContent {
    MessageContent::Leaf(LeafContent::Image {
        media_type: media_type.to_string(),
        data: data.to_string(),
        parent_tool_use_id: None,
    })
}

pub(crate) fn tool_use_block(
    id: &str,
    name: &str,
    input: HashMap<String, Value>,
) -> MessageContent {
    MessageContent::Node(mainframe_types::chat::MessageContentNode::ToolUse {
        id: id.to_string(),
        name: name.to_string(),
        input,
        parent_tool_use_id: None,
    })
}

pub(crate) fn tool_result_block(
    tool_use_id: &str,
    content: &str,
    is_error: bool,
    structured_patch: Option<Vec<DiffHunk>>,
) -> MessageContent {
    MessageContent::Node(mainframe_types::chat::MessageContentNode::ToolResult {
        tool_use_id: tool_use_id.to_string(),
        content: content.to_string(),
        is_error,
        structured_patch,
        original_file: None,
        modified_file: None,
        parent_tool_use_id: None,
    })
}

pub(crate) fn collab_agent_tool_use(
    id: &str,
    prompt: &str,
    description: &str,
    subagent_type: &str,
) -> MessageContent {
    let mut input = HashMap::new();
    input.insert("prompt".to_string(), json!(prompt));
    input.insert("description".to_string(), json!(description));
    input.insert("subagent_type".to_string(), json!(subagent_type));
    tool_use_block(id, "CollabAgent", input)
}

/// Tag a block with `parentToolUseId` (mirrors the TS `{ ...b, parentToolUseId }`).
pub(crate) fn with_parent(block: MessageContent, pid: &str) -> MessageContent {
    use mainframe_types::chat::MessageContentNode as N;
    let pid = Some(pid.to_string());
    match block {
        MessageContent::Leaf(LeafContent::Text { text, .. }) => {
            MessageContent::Leaf(LeafContent::Text {
                text,
                parent_tool_use_id: pid,
            })
        }
        MessageContent::Leaf(LeafContent::Thinking { thinking, .. }) => {
            MessageContent::Leaf(LeafContent::Thinking {
                thinking,
                parent_tool_use_id: pid,
            })
        }
        MessageContent::Leaf(LeafContent::Image {
            media_type, data, ..
        }) => MessageContent::Leaf(LeafContent::Image {
            media_type,
            data,
            parent_tool_use_id: pid,
        }),
        MessageContent::Leaf(LeafContent::SkillLoaded {
            skill_name,
            path,
            content,
            ..
        }) => MessageContent::Leaf(LeafContent::SkillLoaded {
            skill_name,
            path,
            content,
            parent_tool_use_id: pid,
        }),
        MessageContent::Node(N::ToolUse {
            id, name, input, ..
        }) => MessageContent::Node(N::ToolUse {
            id,
            name,
            input,
            parent_tool_use_id: pid,
        }),
        MessageContent::Node(N::ToolResult {
            tool_use_id,
            content,
            is_error,
            structured_patch,
            original_file,
            modified_file,
            ..
        }) => MessageContent::Node(N::ToolResult {
            tool_use_id,
            content,
            is_error,
            structured_patch,
            original_file,
            modified_file,
            parent_tool_use_id: pid,
        }),
        MessageContent::Node(N::PermissionRequest { request, .. }) => {
            MessageContent::Node(N::PermissionRequest {
                request,
                parent_tool_use_id: pid,
            })
        }
        MessageContent::Node(N::Error { message, .. }) => MessageContent::Node(N::Error {
            message,
            parent_tool_use_id: pid,
        }),
        MessageContent::Node(N::Compaction { .. }) => MessageContent::Node(N::Compaction {
            parent_tool_use_id: pid,
        }),
    }
}

// PORT STATUS: src/plugins/builtin/codex/history.ts (249 lines)
// confidence: medium
// todos: 1
// notes: BLOCKER — `parse_unified_diff` (now in unified_diff.rs) lives in a
// notes: crate-private shim (faithful copy of messages/parse-unified-diff.ts)
// notes: because mainframe_display::parse_unified_diff is still a skeleton; swap to
// notes: the canonical fn once that task lands (TODO(port)).
// notes: This file hosts the crate-shared MessageContent block builders + the
// notes: `with_parent` tagger (reused by event_mapper) to keep ONE canonical copy.
// notes: convert_thread_items (now in history_convert.rs) takes all 4 params
// notes: explicitly (TS defaulted the last two); the recursive child call passes an
// notes: empty agent-meta map, matching TS.
// notes: Tests in tests/history.rs — both codex/__tests__/history.test.ts (userMessage
// notes: shapes + id stability) AND src/__tests__/codex-history.test.ts (per-item-type
// notes: conversions), assertion-for-assertion.
// notes: task 2 (todo #247) carved convert_thread_items into history_convert.rs,
// notes: emit_collab_agent into history_collab.rs, and parse_unified_diff/
// notes: parse_hunk_header/parse_pair into unified_diff.rs; re-exported here so
// notes: external `history::X` call sites keep compiling.
